query_execution/Foreman.cpp

```diff
@@ -192,15 +182,56 @@ void Foreman::run() {
   while (!checkQueryExecutionFinished()) {
     // Receive() causes this thread to sleep until next message is received.
     AnnotatedMessage annotated_msg = bus_->Receive(foreman_client_id_, 0, true);
-    // Message is either workorder feedback message or foreman message.
-    if (annotated_msg.tagged_message.message_type() == kWorkOrderFeedbackMessage) {
-      WorkOrder::FeedbackMessage msg(
-          const_cast<void *>(annotated_msg.tagged_message.message()),
-          annotated_msg.tagged_message.message_bytes());
-      processFeedbackMessage(msg);
-    } else {
-      processMessage(*static_cast<const ForemanMessage *>(
-          annotated_msg.tagged_message.message()));
+    const TaggedMessage &tagged_message = annotated_msg.tagged_message;
+    switch (tagged_message.message_type()) {
+      case kWorkOrderCompleteMessage: {
+        serialization::WorkOrderCompletionMessage proto;
+        CHECK(proto.ParseFromArray(tagged_message.message(), tagged_message.message_bytes()));
+
+        processWorkOrderCompleteMessage(proto.operator_index(), proto.worker_id());
+        break;
+      }
+      case kRebuildWorkOrderCompleteMessage: {
+        serialization::WorkOrderCompletionMessage proto;
+        CHECK(proto.ParseFromArray(tagged_message.message(), tagged_message.message_bytes()));
+
+        processRebuildWorkOrderCompleteMessage(proto.operator_index(), proto.worker_id());
+        break;
+      }
+      case kDataPipelineMessage: {
+        // Possible message senders include InsertDestinations and some
+        // operators which modify existing blocks.
+        serialization::DataPipelineMessage proto;
+        CHECK(proto.ParseFromArray(tagged_message.message(), tagged_message.message_bytes()));
+
+        processDataPipelineMessage(proto.operator_index(), proto.block_id(), proto.relation_id());
+        break;
+      }
+      case kWorkOrdersAvailableMessage: {
+        serialization::WorkOrdersAvailableMessage proto;
+        CHECK(proto.ParseFromArray(tagged_message.message(), tagged_message.message_bytes()));
+
+        const dag_node_index op_index = proto.operator_index();
+
+        // Check if new work orders are available.
+        fetchNormalWorkOrders(op_index);
+
+        // Dispatch the WorkerMessages to the workers. We prefer to start the search
+        // for the schedulable WorkOrders beginning from 'op_index'. The first
+        // candidate worker to receive the next WorkOrder is the one that sent the
+        // response message to Foreman.
+        // TODO(zuyu): Improve the data locality for the next WorkOrder.
+        dispatchWorkerMessages(0, op_index);
+        break;
+      }
+      case kWorkOrderFeedbackMessage: {
+        WorkOrder::FeedbackMessage msg(const_cast<void *>(tagged_message.message()),
```

Comment:
The answer is yes and no.

For yes part, so far we do not create a proto for the `feedback message`, and it works fine.

On the other side, for now every discrete `WorkOrder feedback message` is a deserialized proto (e.g., in `TextScanOp` and `Sort`-related operators), and it is up to the operators to interpret the underlying meanings of the feedback message in `receiveFeedbackMessage`.